server/neptune/workflows/internal/deploy/workflow.go

```diff
@@ -0,0 +1,129 @@
+package deploy
+
+import (
+	"time"
+
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config/logger"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/revision/queue"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/signals"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/github"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	TaskQueue = "deploy"
+)
+
+type TimedReceiver interface {
+	DidTimeout() bool
+	AddTimeout(ctx workflow.Context, selector workflow.Selector)
+}
+
+type QueueWorker interface {
+	Work(ctx workflow.Context)
+	GetState() queue.WorkerState
+}
+
+func Workflow(ctx workflow.Context, request Request) error {
+	options := workflow.ActivityOptions{
+		TaskQueue:              TaskQueue,
+		ScheduleToCloseTimeout: 5 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, options)
+
+	runner := newRunner(ctx, request)
+
+	// blocking call
+	return runner.Run(ctx)
+}
+
+type Runner struct {
+	QueueWorker      QueueWorker
+	Selector         workflow.Selector
+	RevisionReceiver TimedReceiver
+}
+
+func newRunner(ctx workflow.Context, request Request) *Runner {
+	// set relevant logging context
+	ctx = workflow.WithValue(ctx, config.RepositoryLogKey, request.Repository.FullName)
+	ctx = workflow.WithValue(ctx, config.ProjectLogKey, request.Root.Name)
+	ctx = workflow.WithValue(ctx, config.GHRequestIDLogKey, request.GHRequestID)
+
+	// convert to internal types, we should probably move these into another struct
+	repo := github.Repo{
+		Name:     request.Repository.Name,
+		Owner:    request.Repository.Owner,
+		FullName: request.Repository.FullName,
+		URL:      request.Repository.URL,
+	}
+
+	// inject dependencies
+
+	// temporal effectively "injects" this, it just cares about the method names,
+	// so we're modeling our own DI around this.
+	var a *activities.Deploy
+
+	revisionQueue := queue.NewQueue()
+	revisionReceiver := signals.NewRevisionSignalReceiver(ctx, revisionQueue, 60*time.Second)
+	selector := workflow.NewSelector(ctx)
+	revisionReceiver.AddCallback(ctx, selector)
+
+	worker := &queue.Worker{
+		Queue:      revisionQueue,
+		Activities: a,
+		Repo:       repo,
+		RootName:   request.Root.Name,
+	}
+
+	return &Runner{
+		QueueWorker:      worker,
+		Selector:         selector,
+		RevisionReceiver: revisionReceiver,
+	}
+}
+
+func (r *Runner) Run(ctx workflow.Context) error {
+	workerCtx, cancel := workflow.WithCancel(ctx)
+
+	wg := workflow.NewWaitGroup(ctx)
+	wg.Add(1)
+
+	// if this panics in anyway, we'll need to ship a fix to the running workflows, else risk dropping
+	// signals
+	// should we have some way of persisting our queue in case of workflow termination?
```

Comment:
doesn't workflows end up creating a queue?

Reply:
noo they don't a workflow is just running code serially.  In other cases, we've emulated a queue by using a for select.  In this case, we need visibility into the queue items to do things like notifying when a revision is queued/locked etc.